sets.c

```diff
@@ -0,0 +1,215 @@
+            if (errno || *tmpnum_endptr != '\0')
+                goto error;
+
+            free((void *)tmpnum);
+            tmpnum = NULL;
+
+            if (last_type == RANGE_TOK) {
+                if (!got_num) {
+                    /* we've got a {1 .. n} range */
+                    last_num  = toknum;
+                    last_type = NUMBER_TOK;
+                    got_num   = true;
+                    /* move token position to last number found (needed for 2+ digit nums */
+                    tok = tmpnum_end-1;
+                    /* bounds check */
+                    if (last_num >= maxsz)
+                        goto error;
+                    /* pre-set the new size filled and grow array if necessary */
+                    if ((arr_szfilled += last_num) > arr_maxsz)
+                        while (arr_szfilled > arr_maxsz)
+                            INC_ARRAY_SZ;
+                    /* fill up array with range */
+                    for (size_t i = 0; i <= last_num; i++)
+                        valarr[vaidx++] = i;
+                    continue;
+                }
+                if (toknum <= last_num || toknum >= maxsz)
+                    goto error; /* invalid range */
+
+                /* pre-set the new size filled and grow array if necessary */
+                if ((arr_szfilled += (toknum - last_num)) > arr_maxsz)
+                    while (arr_szfilled > arr_maxsz)
+                        INC_ARRAY_SZ;
+
+                /* fill up array with range */
+                for (size_t i = last_num+1; i <= toknum; i++)
+                    valarr[vaidx++] = i;
+            } else if (last_type == NUMBER_TOK) {
+                /* sanity check */
+                goto error;
+            } else {
+                valarr[vaidx++] = toknum;
+                arr_szfilled++;
+            }
+            last_num  = toknum;
+            last_type = NUMBER_TOK;
+            if (!got_num)
+                got_num = true;
+            /* move token position to last number found (needed for 2+ digit nums) */
+            tok = tmpnum_end-1;
+        } else {
+            /* bad token */
+            goto error;
+        }
+    }
+
+    /* check for {n .. end} range */
+    if (last_type == RANGE_TOK) {
+        if (!got_num)
+            goto error;
+        if (last_num >= maxsz)
+            goto error;
+        if ((arr_szfilled += (maxsz-1 - last_num)) > arr_maxsz)
+            while (arr_szfilled > arr_maxsz)
+                INC_ARRAY_SZ;
+        /* fill up array with range */
+        for (size_t i = last_num+1; i < maxsz; i++)
+            valarr[vaidx++] = i;
+    }
+
+    /* consider empty sets invalid */
+    if (arr_szfilled == 0)
+        goto error;
+
+    /* sort the value array */
+    qsort(valarr, arr_szfilled, sizeof(valarr[0]), _size_t_cmp);
+
+    /* check if there are any duplicates */
+    for (size_t i = 0; i < (arr_szfilled % 2 ? arr_szfilled-1 : arr_szfilled); i++)
```

Comment:
This guard is probably wrong, should be `i+1 < arr_szfilled`

Reply:
It works, but this way is indeed cleaner (although an additional add op is introduced, it is negligible ;-)).  Will update

Reply:
Thanks, a couple possible extra uops on a non-critical functions are worth it to avoid confusion every time I see this guard.